libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt

```diff
@@ -1,360 +1,38 @@
-                }
-            }.awaitAll()
-        }
-    }
-
-    private fun logRoutesDiff(
-        newRoute: NavigationRoute,
-        oldRoute: NavigationRoute,
-        currentLegIndex: Int,
-    ) {
-        val routeDiffs = routeDiffProvider.buildRouteDiffs(
-            oldRoute,
-            newRoute,
-            currentLegIndex,
-        )
-        if (routeDiffs.isEmpty()) {
-            logI("No changes in annotations for route ${newRoute.id}", LOG_CATEGORY)
-        } else {
-            for (diff in routeDiffs) {
-                logI(diff, LOG_CATEGORY)
-            }
-        }
-    }
-
-    private suspend fun requestRouteRefresh(
-        route: NavigationRoute,
-        routeRefreshRequestData: RouteRefreshRequestData
-    ): RouteRefreshResult =
-        suspendCancellableCoroutine { continuation ->
-            val requestId = routeRefresh.requestRouteRefresh(
-                route,
-                routeRefreshRequestData,
-                object : NavigationRouterRefreshCallback {
-                    override fun onRefreshReady(route: NavigationRoute) {
-                        continuation.resume(RouteRefreshResult.Success(route))
-                    }
-
-                    override fun onFailure(error: NavigationRouterRefreshError) {
-                        continuation.resume(RouteRefreshResult.Fail(error))
-                    }
-                }
-            )
-            continuation.invokeOnCancellation {
-                routeRefresh.cancelRouteRefreshRequest(requestId)
-            }
-        }
-
-    private suspend fun <T> waitForever(message: String): T {
-        logI("Route won't be refreshed because $message", LOG_CATEGORY)
-        return CompletableDeferred<T>().await()
-    }
-
-    private fun validateRoute(route: NavigationRoute): RouteValidationResult = when {
-        route.routeOptions.enableRefresh() != true ->
-            RouteValidationResult.Invalid("RouteOptions#enableRefresh is false")
-        route.directionsRoute.requestUuid()?.isNotBlank() != true ->
-            RouteValidationResult.Invalid(
-                "DirectionsRoute#requestUuid is blank. " +
-                    "This can be caused by a route being generated by " +
-                    "an Onboard router (in offline mode). " +
-                    "Make sure to switch to an Offboard route when possible, " +
-                    "only Offboard routes support the refresh feature."
-            )
-        else -> RouteValidationResult.Valid
-    }
-
-    private sealed class RouteValidationResult {
-        object Valid : RouteValidationResult()
-        data class Invalid(val reason: String) : RouteValidationResult()
-    }
-
-    private sealed class RouteRefreshResult {
-        data class Success(val route: NavigationRoute) : RouteRefreshResult()
-        data class Fail(val error: NavigationRouterRefreshError) : RouteRefreshResult()
-    }
+@ExperimentalPreviewMapboxNavigationAPI
+interface RouteRefreshController {
```

Comment:
Adding new functions to an `interface` is a breaking change. Are you sure we won't add any after the interface is no longer experimental?

Reply:
Oh shoot, you're right.
I didn't design this interface to be implemented by customer. So it doesn't look like a pure breaking change to me, although technically it is.
It can be easily fixed by just exposing a class, only marking some of the methods public. But I'd like to know your opinion whether it's really an issue. If the interface is not supposed to be implemented, should adding new methods to it be considered a breaking change? I just don't want us to be artificially limiting ourselves. Exposing some API through an interface instead of a class is generally a good approach, but we wouldn't be able to use it if we can't add new methods. cc @mapbox/navigation-android

Reply:
What is wrong with using a class? Also we can consider making the interface sealed, this way Kotlin developers will not be able to implement it.

Reply:
There is nothing wrong. It's just that hiding an implementation behind an interface is good because it doesn't expose any unnecessary details.
I'm just trying to look from the fresh perspective. Is it really a breaking change? Can we use interfaces as API or not?
Sealed interface only solves the problem for kotlin. Implementing it from Java is allowed: there'll be a warning, but it'll compile.

Reply:
Alright, I exposed a class instead.